app/uk/gov/hmrc/agentsubscription/service/SubscriptionService.scala

```diff
@@ -70,19 +72,30 @@ class SubscriptionService @Inject() (
   def subscribeAgentToMtd(subscriptionRequest: SubscriptionRequest, authIds: AuthIds)(implicit hc: HeaderCarrier, ec: ExecutionContext, request: Request[Any]): Future[Option[Arn]] = {
 
     desConnector.getRegistration(subscriptionRequest.utr) flatMap {
-        case Some(DesRegistrationResponse(Some(desPostcode), _, _, _, _))
+        case Some(DesRegistrationResponse(Some(desPostcode), isAnAsAgent, _, _, maybeArn))
           if postcodesMatch(desPostcode, subscriptionRequest.knownFacts.postcode) => {
-          subscribe(subscriptionRequest, authIds)
+          subscribe(subscriptionRequest, authIds, isAnAsAgent, maybeArn)
         }
         case _ =>  Future successful None
     }
   }
 
-  private def subscribe(subscriptionRequest: SubscriptionRequest, authIds: AuthIds)(implicit hc: HeaderCarrier, ec: ExecutionContext, request: Request[Any]): Future[Option[Arn]] = {
-    for {
-      arn <- desConnector.subscribeToAgentServices(subscriptionRequest.utr, desRequest(subscriptionRequest))
+  private def subscribe(subscriptionRequest: SubscriptionRequest, authIds: AuthIds, isAnAsAgent: Boolean, maybeArn: Option[Arn])
+                       (implicit hc: HeaderCarrier, ec: ExecutionContext, request: Request[Any]): Future[Option[Arn]] = {
+
+    def subscribeOps(arn: Arn): Future[Unit] = for {
+      _ <- deleteKnownFacts(arn, subscriptionRequest, authIds)
       _ <- createKnownFacts(arn, subscriptionRequest, authIds)
       _ <- enrol(arn, subscriptionRequest, authIds)
+    } yield ()
+
+    for {
+      arn <- maybeArn match {
+        case Some(arn) if isAnAsAgent => Future.successful(arn)
+        case _ => desConnector.subscribeToAgentServices(subscriptionRequest.utr, desRequest(subscriptionRequest))
+      }
+      alreadyEnrolled <- hasPrincipalGroupIds(arn, subscriptionRequest, authIds)
+      _ <-  if (!alreadyEnrolled) subscribeOps(arn) else Future.failed(EnrolmentAlreadyAllocated("HMRC-AS-AGENT already allocated to a group for this Arn"))
```

Comment:
The message "An enrolment for HMRC-AS-AGENT with this ARN as an identifier already exists" would be better here.